Scheduler that manually advances time

raboof opened the pull request:
#24150

First version for early review. More tests, docs and Java examples still TBD.

raboof committed:
Scheduler that manually advances time (#24150)

First version for review. More tests, docs and Java examples still TBD.

chbatey commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/ManualTimerSpec.scala:
```diff
@@ -0,0 +1,37 @@
+
+class ManualTimerSpec extends TestKit() with ManualTime with WordSpecLike {
+  //#manual-scheduling-simple
+  "A timer" must {
```
Looks nice

johanandren commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,86 @@
+ * For testing: scheduler that does not look at the clock, but must be progressed manually by calling `timePasses`.
+ *
+ * This is not entirely realistic: jobs will be executed on the test thread instead of using the `ExecutionContext`, but does
+ * allow for faster and less timing-sensitive specs..
```
Couldn't we still have them run on whatever execution context they want, but block `time Passes` on the futures completion?

raboof replied:
I'll give that a try!

raboof committed:
Don't apply dilation to scheduler parameter

raboof committed:
Clarify ExecutionContext usage

johanandren commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -38,12 +38,12 @@ class ExplicitlyTriggeredScheduler(config: Config, log: LoggingAdapter, tf: Thre
+   * We will not add a dilation factor to this amount, since the scheduler API also does not apply dilation.
+   * If you want the amount of time passed to be dilated, apply the dilation before passing the delay to
+   * this method.
+   */
```
"This will execute all outstanding jobs on the calling thread before returning" or such?

raboof replied:
👍. (It will also execute any jobs that are scheduled to be executed within the specified period by those outstanding jobs, but that might be a subtlety that doesn't really belong in the scaladoc here)

raboof committed:
Clarify comment on timePasses

johanandren commented on akka-testkit-typed/src/main/scala/akka/testkit/typed/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -11,7 +11,7 @@ import scala.concurrent.duration.{ Duration, FiniteDuration }
 
   @varargs
-  def expectNoMessageFor[T](duration: FiniteDuration, on: scaladsl.TestProbe[T]*): Unit = {
+  def expectNoMessageFor(duration: FiniteDuration, on: scaladsl.TestProbe[_]*): Unit = {
```
Aah, I misread this to be on the probe, which seemed to magic, this is explicit enough for me! :) 👍

johanandren reviewed (approved):
Nice!

patriknw commented on akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java:
```diff
@@ -0,0 +1,54 @@
+
+    spawn(behavior);
+
+    scheduler.expectNoMessageFor(Duration.create(9, TimeUnit.MILLISECONDS), probe);
```
would it be more readable to put the `expectNoMessageFor` on the `probe`, with a `ExplicitlyTriggeredScheduler` parameter?

raboof replied:
I wanted to avoid confusion with `probe.expectNoMessage`, which will not automatically advance the time of the explicit scheduler. If we decide we want to have separate probes for normal vs manual schedulers anyway (see below) then it should indeed of course move to the probe.

patriknw commented on akka-testkit-typed/src/main/scala/akka/testkit/typed/TestKit.scala:
```diff
@@ -64,12 +64,16 @@ class TestKit(name: String, config: Option[Config]) extends TestKitBase {
   import TestKit._
-  implicit val system = ActorSystem(testKitGuardian, name, config = config)
+  implicit val system = ActorSystem(testKitGuardian, name, config = Some(config match {
+    case None                        ⇒ mixedInConfig.withFallback(ConfigFactory.load())
```
I don't think the `withFallback` is needed here, the ActorSystem takes care of that

raboof replied:
AFAICS it is (https://github.com/akka/akka/blob/master/akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala#L176) - which is actually nice for testing, as it opens up the possibility of starting an ActorSystem that is not influenced by any `application.conf` on the classpath at all.

patriknw replied:
yes, application.conf should not be loaded unless no config at all is specified when creating AS

raboof replied:
Right: so this approach now gives you 2 options when 'mixing in' configuration:

* If you want the effective config to be `mixedInConfig.withFallback(ConfigFactory.load())`, do not provide any configuration to `TestKit`
* If you want the effective config to be `explicitConfig.withFallback(mixedInConfig)`, provide the `explicitConfig` as the `TestKit` configuration

When you want the effective config to be something like `explicitConfig.withFallback(mixedInConfig).withFallback(ConfigFactory.load())`, you will have to provide it explicitly as `explicitConfig.withFallback(ManualTime.config).withFallback(ConfigFactory.load())`.

It does make the 'common case' (i.e. `ManualTimerSpec`) nice and simple, but the infrastructure is a bit complicated - hard to say if it is really worth it. I could move the 'mixedInConfig' logic out of this PR and to a separate one?

jrudolph replied:
> It does make the 'common case' (i.e. ManualTimerSpec) nice and simple, but the infrastructure is a bit complicated - hard to say if it is really worth it. I could move the 'mixedInConfig' logic out of this PR and to a separate one?

Maybe splitting off would be a good idea for now. The idea is good, though. It's worthwhile to introduce a principled way of stacking predefined settings (I guess one problem that will remain is that we cannot easily detect conflicts between different mixed-in config but that might be acceptable).

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+  case class Item(time: Long, interval: Option[FiniteDuration], runnable: Runnable)
+
+  val currentTime = new AtomicLong()
+  val scheduled = new ConcurrentHashMap[Item, Unit]()
```
too many things public here?

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+   * this method.
+   */
+  def timePasses(amount: FiniteDuration) = {
+    // Give dispatchers time to clear :(
```
when is that needed and what will happen if they are not done within 10 ms?

raboof replied:
It is needed when:
* the test triggers some asynchronous operations
* those asynchronous operations schedule tasks
* immediately after triggering the asynchronous operations, we manually advance the scheduler to a point in time after the deadline for the scheduled tasks

This is a race condition: when the scheduler queue is checked before the tasks are scheduled, they are (obviously) missed and won't be evaluated.

The 'proper' solution might be to introduce methods like `expectMsg` and friends that are aware of the manually triggered scheduler and can retry checking the task queue before failing. That would mean we re-use less of the `TestProbe` implementation, perhaps we should even have  separate `TestProbe` implementations, for use with 'normal' schedulers and one for use with the explicit scheduler...

raboof replied:
Alternatively, we could consider taking control over the dispatcher as well (perhaps for now just allowing inspection), so we can observe whether it is finished. Could be interesting but would also mean a bit of extra code - WDYT?

patriknw replied:
I'm not quite sure what kind of rabbit whole this is. I'll probably prefer something simple until we get a feeling of what kind of problems/limitations there are. Perhaps @jrudolph has any thoughts.

jrudolph replied:
> Alternatively, we could consider taking control over the dispatcher as well

That's something that I would have found useful quite a few times as well. The question is whether and how we could reliably determine that a dispatcher is done for now? We could say it's good enough if the task queue is empty and no threads are running (ignoring external task submissions). But even that might be hard to check atomically because executors tend to be optimized not to need a single lock.

I guess we could define a dispatcher wrapper for testing that would keep track of all task submission before submitting to the real dispatcher counting an atomic long up and down?

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+    log.debug(s"Time proceeds from ${currentTime.get} to $newTime, currently scheduled for this period:")
+    scheduledTasks(newTime).foreach(item ⇒
+      log.debug(s"- $item")
+    )
```
construct one single String and `log.debug` that instead, this will result in many independent log lines

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+        currentTime.set(task.time)
+        task.runnable.run()
+        scheduled.remove(task)
+        task.interval.foreach(i ⇒ scheduled.put(task.copy(time = task.time + i.toMillis), ()))
```
look at how the real scheduler handles exceptions thrown from `run`. I think they should be caught and periodic tasks cancelled.

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+  private def schedule(initialDelay: FiniteDuration, interval: Option[FiniteDuration], runnable: Runnable): Cancellable = {
+    val firstTime = currentTime.get + initialDelay.toMillis
+    val item = Item(firstTime, interval, runnable)
+    log.debug(s"Scheduled item for $firstTime: $item")
```
use param placeholders, `{}`

patriknw commented on akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala:
```diff
@@ -0,0 +1,107 @@
+    log.debug(s"Scheduled item for $firstTime: $item")
+    scheduled.put(item, ())
+
+    if (initialDelay == Duration.Zero)
```
perhaps `<=` ?

patriknw reviewed (commented):
very useful, a few small things

raboof committed:
Make ExplicitlyTriggeredScheduler internals private

raboof committed:
List currently scheduled tasks in one log message

raboof committed:
Execute immediately if (initialDelay <= Duration.Zero)

raboof committed:
Don't reschedule if scheduled task fails

raboof committed:
Be more efficient about logging

raboof committed:
Widen `timePasses` delay for now

https://github.com/akka/akka/pull/24243#discussion_r160985493 for some
discussion on what to do instead

raboof committed:
Remove mechanism for mixing in config from a test trait

patriknw reviewed (approved):
LGTM
